examples/federated_learning_with_encryption.py

```diff
@@ -0,0 +1,234 @@
+"""
+In this example we assume to deal with sensitive data of 442 hospital patients,
+with different level of progress of diabetes. Recorded variables are age,
+gender, body mass index, average blood pressure, and six blood serum
+measurements. A last variable is a quantitative measure of the disease
+progression. Since this measure is continuous, we will solve the problem by
+linear regression.
+
+The data is distributed among 3 hospitals, referred as `clients`. The objective
+is to make use of the whole (virtual) training set to improve upon the
+model that can be trained locally. 50 patients will be kept as a testset
+and not used for training. An additional agent is the
+`server`, who will facilitate the information exchange among the hospitals
+under the following constraints. Due to privacy policy:
+
+1) The individual patients' record at each hospital cannot leave its premises,
+not even in encrypted form
+2) Even aggregated information/summary (read: gradients) of the individual
+records cannot leave the hospitals, unless they are first encrypted
+3) None of the parties (clients AND server) must be able to infer WHERE
+(in which hospital) a patient in the training set has been treated.
+
+We solve linear regression by gradient descent. The server owns the private
+key and the clients own the public key. The protocol works as follows.
+Until convergence: hospital 1 computes its gradient, encrypts it and sends it
+to hospital 2; hospital 2 computes its gradient, encrypts and sums it to
+hospital 1's; hospital 3 does the same and passes the overall sum to the
+server. The server obtains the gradient of the whole (virtual)
+training set; it decrypts it and sends it back in the clear to every client,
+who can update the respective local models.
```

Comment:
There is an apparent contradiction here: In 2) above it says that aggregated information cannot leave the hospitals unless it's encrypted, but here it says that the server decrypts the aggregated information and sends it to the clients in the clear. The gradient information sent back to the clients is perhaps "safe" because it has been added to the gradients of the other clients; if so, this should be mentioned explicitly.